psi4/driver/driver_findif.py

```diff
@@ -0,0 +1,812 @@
+    }
+
+    try:
+        print_msg = msg_dict[mode]
+    except KeyError:
+        raise ValidationError("FINDIF: Mode {} not recognized.".format(mode))
+
+    def init_string(data):
+        return ("  Using finite-differences of {:s}.\n"
+                "\tGenerating geometries for use with {:d}-point formula.\n"
+                "\tDisplacement size will be {:6.2e}.\n".format(print_msg, data["num_pts"], data["disp_size"]))
+
+    data = _initialize_findif(mol, freq_irrep_only, mode, init_string, 1)
+
+    # We can finally start generating displacements.
+    ref_geom_temp = mol.geometry()
+    ref_geom = ref_geom_temp.clone()
+    ref_geom.name = "Reference geometry"
+
+    disp_geoms = []
+
+    def append_geoms(indices, steps):
+        """Given a list of indices and a list of steps to displace each, append the corresponding geometry to the list."""
+        new_geom = ref_geom.clone()
+        # Next, to make this salc/magnitude composite.
+        index_steps = zip(indices, steps)
+        _displace_cart(mol, new_geom, data["salc_list"], index_steps, data["disp_size"])
+        disp_geoms.append(new_geom)
+
+    for h in range(data["Nirrep"]):
+        active_indices = data["salc_indices_pi"][h]
+
+        for index in active_indices:
+            # Displace along the diagonal.
+            # Remember that the totally symmetric irrep has special displacements.
+            for val in data["disps"]["sym" if h == 0 else "asym"]:
+                append_geoms((index, ), val)
+
+        # Hessian from energies? We have off-diagonal displacements to worry about.
+        if mode == "2_0":
+            # i indexes SALC indices of the current irrep.
+            for i, index in enumerate(active_indices):
+                for index2 in active_indices[:i]:
+                    for val in data["disps"]["off"]:
+                        append_geoms((index, index2), val)
+
+    disp_geoms.append(ref_geom)
+
+    if data["print_lvl"] > 2:
+        for disp in disp_geoms:
+            disp.print_out()
+
+    if data["print_lvl"] > 1:
+        core.print_out("\n-------------------------------------------------------------\n")
+
+    return disp_geoms
+
+
+def comp_grad_from_energy(mol, E):
+    """Compute the gradient by finite difference of energies.
+
+    Parameters
+    ----------
+    mol : qcdb.molecule or psi4.core.Molecule
+        The molecule to compute the gradient of.
+    E : list of floats
+        A list of energies of the molecule at displaced geometries.
+
+    Returns
+    -------
+    gradient : psi4.core.Matrix
+        The gradient in Cartesians, as a matrix with dimensions
+        number-of-atoms by 3. """
+
+    # NOTE: Yes, this returns psi4.core.Matrix, not an ndarray.
+    # Due to GradientWriter, the function internally needs the gradient as a psi4.core.Matrix.
+    # TODO: Move GradientWriter py-side, as well? This function would simplify nicely.
```

Comment:
How difficult would it be to move GradientWriter py-side, too? I'm not too familiar with where it is C-side.

Reply:
It doesn't look hard at all. GradientWriter is defined in `libmints/writer.*` and only ever used in `findif/fd_1_0.cc`. I don't know how we write analytic gradients (would have to check), but if the grep on `GradientWriter` didn't catch anything else, it's probably Python tech that can be reused for findif.

Reply:
Then I say let's move it py-side.

Reply:
Shall I write it as a qcdb module, or a standalone in the driver? We'd be fine with a standalone for now, but I can very easily envision somebody wanting a load_gradient option. Maybe they optimized in one job and want to skip the initial gradient computation when doing a hessian in another. If we have writing and reading gradients, it would make more sense to have a module for the both of them, just like we do with hessians.

Reply:
I think it makes sense to parallel hessian as much as possible, so add to qcdb and give the format a name so the fun can handle variations in future.

Reply:
Next question for the core devs: The gradients are currently printed in file11 format, where the first line is the molecule name, the word "(wfn)" and the word "(dertype)". This is followed by the number of atoms and current energy, the masses and coordinates, and _then_ the gradient.

Keep all the extra stuff, or prune the gradient file to be an analogue of the hessian file? @jturney couldn't quite decide when I asked him.

Reply:
Oh, I just realized you were debating whether to print the geometry too.  I will just say that in practice I find myself working with, e.g., CFOUR hessian file (with which I am less familiar) and wondering if I am completely sure what Cartesian geometry it belongs to.  Perhaps with modern object packaging of molecular data this problem will go away, but it's nice to be able to verify.

Reply:
Another picky question: We currently write multiple gradients to the same file. See fd-gradient test for an example. Keep this behavior or scrap it? My vote would be to have one gradient for one file, but there may be some use case I'm not seeing.